### PR TITLE
Benchmark scripts for attn_backend

### PR DESCRIPTION
**Pull request by byjiang1996**
Author: shivam15s <shivam15800@gmail.com>
Author: Vaibhav Jindal <32337828+vaibhavjindal@users.noreply.github.com>

## Motivation

Benchmark Attention Backends with different features and models for issue https://github.com/sgl-project/sglang/issues/5172

## Modifications

```
python bench.py --config bench_serving.yaml bench_serving_spec_decode.yaml gsm8k.yaml gsm8k_spec_decode.yaml 
```
Benchmark results will be stored in `<torch.cuda.get_device_name(0).replace(' ', '_')>.md` file

### Experiment - bench_serving | DeepSeek-V3

| Configuration | Input token throughput (tok/s) | Output token throughput (tok/s) |
|-------------|----------------|----------------|
| attn=fa3,random_input_len=1024,random_output_len=1024 | 2216.96    | 2268.68    |
| attn=flashinfer,random_input_len=1024,random_output_len=1024 | 2226.81    | 2278.76    |
| attn=fa3,random_input_len=8192,random_output_len=1024 | 8369.41    | 1078.81    |
| attn=flashinfer,random_input_len=8192,random_output_len=1024 | 8373.39    | 1079.32    |
| attn=fa3,random_input_len=8192,random_output_len=8192 | 1385.71    | 1482.43    |
| attn=flashinfer,random_input_len=8192,random_output_len=8192 | 1418.84    | 1517.88    |

### Experiment - bench_serving | Meta-Llama-3.1-8B-Instruct

| Configuration | Input token throughput (tok/s) | Output token throughput (tok/s) |
|-------------|----------------|----------------|
| attn=fa3,random_input_len=1024,random_output_len=1024 | 5232.96    | 5355.04    |
| attn=flashinfer,random_input_len=1024,random_output_len=1024 | 4819.61    | 4932.05    |
| attn=fa3,random_input_len=8192,random_output_len=1024 | 17652.56   | 2275.40    |
| attn=flashinfer,random_input_len=8192,random_output_len=1024 | 13886.84   | 1790.00    |
| attn=fa3,random_input_len=8192,random_output_len=8192 | 2419.37    | 2588.25    |
| attn=flashinfer,random_input_len=8192,random_output_len=8192 | 1641.00    | 1755.54    |

### Experiment - bench_serving_spec_decode | DeepSeek-V3

| Configuration | Input token throughput (tok/s) | Output token throughput (tok/s) |
|-------------|----------------|----------------|
| attn=fa3,random_input_len=1024,random_output_len=1024 | 739.87     | 757.13     |
| attn=flashinfer,random_input_len=1024,random_output_len=1024 | 712.76     | 729.38     |
| attn=fa3,random_input_len=8192,random_output_len=1024 | 4742.48    | 611.30     |
| attn=flashinfer,random_input_len=8192,random_output_len=1024 | 4673.64    | 602.43     |
| attn=fa3,random_input_len=8192,random_output_len=8192 | 862.90     | 923.14     |
| attn=flashinfer,random_input_len=8192,random_output_len=8192 | 870.02     | 930.75     |

### Experiment - bench_serving_spec_decode | Meta-Llama-3.1-8B-Instruct

| Configuration | Input token throughput (tok/s) | Output token throughput (tok/s) |
|-------------|----------------|----------------|
| attn=fa3,random_input_len=1024,random_output_len=1024 | 4500.17    | 4605.15    |
| attn=flashinfer,random_input_len=1024,random_output_len=1024 | 3506.93    | 3588.74    |
| attn=fa3,random_input_len=8192,random_output_len=1024 | 11616.10   | 1497.30    |
| attn=flashinfer,random_input_len=8192,random_output_len=1024 | 6776.90    | 873.53     |
| attn=fa3,random_input_len=8192,random_output_len=8192 | 1425.65    | 1525.17    |
| attn=flashinfer,random_input_len=8192,random_output_len=8192 | 626.60     | 670.34     |

### Experiment - gsm8k | DeepSeek-V3

| Configuration | Accuracy | Output throughput |
|-------------|----------------|----------------|
| attn=fa3 | 0.970 | 573.756 token/s |
| attn=flashinfer | 0.960 | 570.449 token/s |

### Experiment - gsm8k_spec_decode | DeepSeek-V3

| Configuration | Accuracy | Output throughput |
|-------------|----------------|----------------|
| attn=fa3 | 0.960 | 409.504 token/s |
| attn=flashinfer | 0.970 | 390.942 token/s |

### Experiment - gsm8k | Meta-Llama-3.1-8B-Instruct

| Configuration | Accuracy | Output throughput |
|-------------|----------------|----------------|
| attn=fa3 | 0.755 | 3837.773 token/s |
| attn=flashinfer | 0.760 | 3068.449 token/s |

### Experiment - gsm8k_spec_decode | Meta-Llama-3.1-8B-Instruct

| Configuration | Accuracy | Output throughput |
|-------------|----------------|----------------|
| attn=fa3 | 0.775 | 3234.893 token/s |
| attn=flashinfer | 0.775 | 2689.991 token/s |
## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Review by hebiao064 — APPROVED** [review]
LGTM, thanks for the great job!

File: benchmark/attention_backend/NVIDIA_H200.md [hunk 1]
```diff
@@ -0,0 +1,308 @@
+## Experiment - bench_serving | DeepSeek-V3
+
+| Configuration | Input token throughput (tok/s) | Output token throughput (tok/s) |
```
Comment (zhyncs) [inline benchmark/attention_backend/NVIDIA_H200.md:3]: We typically do not include the result in the document.

File: benchmark/attention_backend/NVIDIA_H200.md [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+
+```
+python3 -m sglang.launch_server --attention-backend fa3 --tp 8 --model-path /shared/public/elr-models/deepseek-ai/DeepSeek-V3/1d044fd82b15f1cedb197a288e50cc96a2c27205
+python3 -m sglang.bench_serving --backend sglang --dataset-name random --request-rate 16 --num-prompts 200 --random-input-len 1024 --random-output-len 1024
+```
+
+#### attn=flashinfer,random_input_len=1024,random_output_len=1024
+
+```
+python3 -m sglang.launch_server --attention-backend flashinfer --tp 8 --model-path /shared/public/elr-models/deepseek-ai/DeepSeek-V3/1d044fd82b15f1cedb197a288e50cc96a2c27205
```
Comment (zhyncs) [inline benchmark/attention_backend/NVIDIA_H200.md:24]: Use model id instead of local path.

File: benchmark/attention_backend/NVIDIA_H200.md [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+| attn=fa3,random_input_len=8192,random_output_len=8192 | 1385.71    | 1482.43    |
+| attn=flashinfer,random_input_len=8192,random_output_len=8192 | 1418.84    | 1517.88    |
+
+### Commands
+
+#### attn=fa3,random_input_len=1024,random_output_len=1024
+
+```
+python3 -m sglang.launch_server --attention-backend fa3 --tp 8 --model-path /shared/public/elr-models/deepseek-ai/DeepSeek-V3/1d044fd82b15f1cedb197a288e50cc96a2c27205
+python3 -m sglang.bench_serving --backend sglang --dataset-name random --request-rate 16 --num-prompts 200 --random-input-len 1024 --random-output-len 1024
```
Comment (zhyncs) [inline benchmark/attention_backend/NVIDIA_H200.md:18]: The default random ratio is 0, so the isl/osl is not 1024. The others are similar.

**Review by zhyncs — CHANGES_REQUESTED** [review]

File: benchmark/attention_backend/NVIDIA_H200.md [hunk 4]
```diff
@@ -0,0 +1,308 @@
+## Experiment - bench_serving | DeepSeek-V3
```
Comment (zhyncs) [inline benchmark/attention_backend/NVIDIA_H200.md:1]: BTW When benchmarking the DeepSeek V3, Llama 4, and Qwen 3, we typically use `max-concurrency` instead of `request-rate` for online use cases due to the model's size.

**Review by zhyncs — CHANGES_REQUESTED** [review]

File: benchmark/attention_backend/bench.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+from collections import defaultdict
+from dataclasses import dataclass
+from io import StringIO
+from pathlib import Path
+from typing import List, Tuple
+
+import torch
+import yaml
+
+from sglang.utils import launch_server_cmd, terminate_process, wait_for_server
```
Comment (zhyncs) [inline benchmark/attention_backend/bench.py:15]: Typically, we avoid introducing dependencies related to sglang in order to make the script more standalone and usable without installing sglang.

File: benchmark/attention_backend/bench.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+    return process, output_buffer
+
+
+def run_task(config: TaskConfig) -> TaskResult:
+    try:
+        print(f"Starting server: {config.server_cmd}")
+        server_process, port = launch_server_cmd(config.server_cmd)
+
+        # timeout = 300s
+        wait_for_server(f"http://localhost:{port}", 300)
```
Comment (zhyncs) [inline benchmark/attention_backend/bench.py:85]: Some models may require more than 300 seconds.

File: benchmark/attention_backend/bench.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
+import argparse
```
Comment (zhyncs) [inline benchmark/attention_backend/bench.py:1]: Why not use the code from https://github.com/sgl-project/sglang/blob/main/test/srt/experiment_runner.py? Just need to add support for exporting the results, everything else can be reused.

**Review by zhyncs — CHANGES_REQUESTED** [review]

**Comment by byjiang1996** [comment]
Close the PR b/c most benchmark script logic are duplicated with https://github.com/sgl-project/sglang/blob/main/test/srt/experiment_runner.py
